block/src/raw_sync.rs

```diff
@@ -142,18 +155,31 @@ impl AsyncIo for RawFileSync {
     }
 
     fn punch_hole(&mut self, offset: u64, len: u64, user_data: u64) -> AsyncIoResult<()> {
-        // SAFETY: FFI call with valid arguments
-        let result = unsafe {
-            libc::fallocate(
-                self.fd,
-                libc::FALLOC_FL_PUNCH_HOLE | libc::FALLOC_FL_KEEP_SIZE,
-                offset as libc::off_t,
-                len as libc::off_t,
-            )
+        let result = if is_block_device(self.fd) {
+            let range: [u64; 2] = [offset, len];
+            // SAFETY: FFI call with valid arguments
+            unsafe {
+                ioctl(
+                    self.fd as libc::c_int,
+                    BLKDISCARD() as _,
+                    &range as *const u64,
+                )
+            }
+        } else {
+            // SAFETY: FFI call with valid arguments
+            unsafe {
+                libc::fallocate(
+                    self.fd,
+                    libc::FALLOC_FL_PUNCH_HOLE | libc::FALLOC_FL_KEEP_SIZE,
+                    offset as libc::off_t,
+                    len as libc::off_t,
+                )
+            }
         };
-        if result < 0 {
+        if result != 0 {
```

Comment:
Why did you change this - is the ioctl returning a positive non-zero value on error?